ns for fx: update graph matching to not match nodes with equal types

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52799 ns for fx: remove ".stats" suffix
* #52798 ns for fx: add node name and type to results
* #52789 ns for fx: make return type of ns APIs future proof
* #52779 ns for fx: unify return types of weight and activation APIs
* #52771 ns for fx: decouple subgraph names from node names
* #52681 ns for fx: update graph matching to handle dicts and tuples in node args
* **#52402 ns for fx: update graph matching to not match nodes with equal types**
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat

Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26499848](https://our.internmc.facebook.com/intern/diff/D26499848)

vkuzo committed:
ns for fx: update graph matching to not match nodes with equal types

Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
rebase on "ns for fx: update graph matching to not match nodes with equal types"


Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26499848](https://our.internmc.facebook.com/intern/diff/D26499848)

[ghstack-poisoned]

vkuzo committed:
fix asan on "ns for fx: update graph matching to not match nodes with equal types"


Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26499848](https://our.internmc.facebook.com/intern/diff/D26499848)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -581,8 +581,8 @@ def __init__(self):
             def forward(self, x0):
-                x1 = F.relu(x0)
-                y1 = F.relu(x0)
+                x1 = torch.add(x0, 1.0)
```
Just curious why change relu to add?

vkuzo replied:
the PR changes nodes with equal types to not be matched by default.  Since relu is the same op before and after quantization, it will no longer be matched in this particular test case (since it's not fused to anything).  We change this to an op which does get swapped, just so this test case still makes sense.  Cat is one example, various other ops could have worked too.

hx89 reviewed (approved):
LGTM!

vkuzo committed:
Update on "ns for fx: update graph matching to not match nodes with equal types"


Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26499848](https://our.internmc.facebook.com/intern/diff/D26499848)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: update graph matching to not match nodes with equal types"


Summary:

Before this PR, any pair of subgraphs with base nodes of equal
types matched.

While sometimes this is useful, this should be off by default to
properly handle user defined modules and functions, for which we do not
know how to extract weights or cast to the right input type.

In a future PR, we can add hooks to turn on matching for nodes
of equal types, for the situations where it makes sense.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26499848](https://our.internmc.facebook.com/intern/diff/D26499848)

[ghstack-poisoned]
